src/Agent.Worker/TaskCommandExtension.cs

```diff
@@ -511,6 +515,76 @@ private void ProcessTaskSetTaskVariableCommand(IExecutionContext context, Dictio
             context.TaskVariables.Set(name, data, isSecret);
         }
 
+        private void ProcessTaskSetEndpointCommand(IExecutionContext context, Dictionary<string, string> eventProperties, string data)
+        {
+            if (string.IsNullOrEmpty(data))
+            {
+                throw new Exception(StringUtil.Loc("EnterValidValueFor0", "setendpoint"));
+            }
+
+            String field;
+            if (!eventProperties.TryGetValue(TaskSetEndpointEventProperties.Field, out field) || String.IsNullOrEmpty(field))
+            {
+                throw new Exception(StringUtil.Loc("MissingEndpointField"));
+            }
+
+            // Mask auth parameter data upfront to avoid accidental secret exposure by invalid endpoint/key/data 
+            if (String.Equals(field, "authParameter", StringComparison.OrdinalIgnoreCase))
+            {
+                var _secretMasker = HostContext.GetService<ISecretMasker>();
+                _secretMasker.AddRegex(data);
+            }
+            
+            String endpointIdInput;
+            if (!eventProperties.TryGetValue(TaskSetEndpointEventProperties.EndpointId, out endpointIdInput) || String.IsNullOrEmpty(endpointIdInput))
+            {
+                throw new Exception(StringUtil.Loc("MissingEndpointId"));
+            }
+
+            Guid endpointId;
+            if (!Guid.TryParse(endpointIdInput, out endpointId))
+            {
+                throw new Exception(StringUtil.Loc("InvalidEndpointId"));
+            }
+
+            var endpoint = context.Endpoints.Find(a => a.Id == endpointId);
```

Comment:
what is .Find()? `context.Endpoints` is a list, wouldn't this be FirstOrDefault/SingleOrDefault and throw a meaningful error if not found?

Reply:
Nevermind, it looks like it's effectively the same thing as FirstOrDefault.

Is ServiceEndpoint a struct?

Reply:
ServiceEndpoint is a class:
https://mseng.visualstudio.com/VSOnline/_git/VSO?path=%2FDistributedTask%2FClient%2FWebApi%2FWebApi%2FServiceEndpoint.cs&version=GBmaster&_a=contents&line=13&lineStyle=plain&lineEnd=15&lineStartColumn=1&lineEndColumn=1

Reply:
I got confused by the line `if (EqualityComparer<ServiceEndpoint>.Default.Equals(endpoint, default(ServiceEndpoint))`. I wasn't sure whether it would null ref, or return the intended result when endpoint is null (i.e. true instead of false). `if (endpoint == null)` might have been easier for project maintainers to read and conveyed more information at a glance. Anyway doesn't really matter, just spent some time trying to figure out what the code meant and whether it would produce the intended result.

Reply:
Agree, we can simplify this.